File: 2023/23.py
# https://adventofcode.com/2023
import pathlib
import sys
from collections import deque
sys.setrecursionlimit(1000000)

filepath = pathlib.Path(__file__)
sys.path.append(str(filepath.parent.parent))
from helpers import * 

data_file = filepath.parent.joinpath(filepath.stem + '.dat')


dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

grid = []
with open(data_file) as f:
    for y, line in enumerate(f.read().splitlines()):
        if y == 0:
            start = (line.index('.'), y)
        end = (line.index('.'), y)
        grid.append(list(line))


max_steps = 0
def recurse(pos, steps, visited):
    x, y = pos

    visited.add(pos)

    global max_steps
    max_steps = max(max_steps, steps)

    if pos == end:
        return
    
    for d_pos in dirs:
        dx, dy = d_pos
        new_pos = x + dx, y + dy
        new_x, new_y = new_pos

        if new_x < 0 or new_y < 0 or new_x >= len(grid[0]) or new_y >= len(grid):
            continue

        if grid[new_y][new_x] == '#':
            continue

        if new_pos in visited:
            continue
        
        recurse((x + dx, y + dy), steps + 1, set(visited))

recurse(start, 0, set())
print(max_steps)





# queue = deque([start, set()])

# while queue:
#     longest =



# part 1
# # https://adventofcode.com/2023
# import pathlib
# import sys
# from collections import deque
# sys.setrecursionlimit(1000000)

# filepath = pathlib.Path(__file__)
# sys.path.append(str(filepath.parent.parent))
# from helpers import * 

# data_file = filepath.parent.joinpath(filepath.stem + '.dat')


# dirs = [(1, 0), (-1, 0), (0, 1), (0, -1)]

# allowed = {
#     '^': (0, -1),
#     '>': (1, 0),
#     '<': (-1, 0),
#     'v': (0, 1),
# }


# grid = []
# with open(data_file) as f:
#     for y, line in enumerate(f.read().splitlines()):
#         if y == 0:
#             start = (line.index('.'), y)
#         end = (line.index('.'), y)
#         grid.append(list(line))


# max_steps = 0
# def recurse(pos, steps, visited):
#     x, y = pos

#     visited.add(pos)

#     global max_steps
#     max_steps = max(max_steps, steps)
    
#     for d_pos in dirs:
#         dx, dy = d_pos
#         new_pos = x + dx, y + dy
#         new_x, new_y = new_pos

#         if new_x < 0 or new_y < 0 or new_x >= len(grid[0]) or new_y >= len(grid):
#             continue

#         if grid[new_y][new_x] == '#':
#             continue

#         if new_pos in visited:
#             continue


#         if grid[new_y][new_x] in '<>^v':
#             if d_pos != allowed[grid[new_y][new_x]]:
#                 continue
        
#         recurse((x + dx, y + dy), steps + 1, set(visited))



# recurse(start, 0, set())
# print(max_steps)





# # queue = deque([start, set()])

# # while queue:
# #     longest =

File: 2023/23.cpp
// g++ 23.cpp -std=c++20 -Ofast -o out && ./out

#include <iostream>
#include <fstream>
#include <vector>
#include <sstream>
#include <chrono>
#include <unordered_map>
#include <unordered_set>
#include <deque>
#include <string>
#include <stdint.h>

using namespace std;
using namespace std::chrono;

vector<pair<int, int>> dirs = {
    make_pair(1, 0),
    make_pair(-1, 0), 
    make_pair(0, 1), 
    make_pair(0, -1),
};

struct pair_hash {
    std::size_t operator () (const std::pair<int, int>& p) const {
        return (p.first + p.second) * (p.first + p.second + 1) / 2 + p.first;
    }
};

int64_t nodes = 0;
auto start = high_resolution_clock::now();
int recurse(vector<vector<char>>& grid, pair<int, int> pos, pair<int, int> end, unordered_set<pair<int, int>, pair_hash>& visited, int steps) {
    nodes++;
    if (nodes % 1000000 == 0 && nodes != 0) {
        auto end = high_resolution_clock::now();
        auto duration = duration_cast<milliseconds>(end - start);
        auto rate = nodes / ((float) duration.count() / 1000);
        cout << nodes << " nodes explored. " << duration.count() << " seconds taken so far. " << rate << " nodes per second." << endl;
    }

    if (pos == end) {
        return steps;
    }

    auto&[x, y] = pos;
    visited.emplace(pos);

    int the_max = 0;
    for (auto& dpos: dirs) {
        auto&[dx, dy] = dpos;
        pair<int, int> new_pos = make_pair(x + dx, y + dy);
        auto&[new_x, new_y] = new_pos;

        if (new_x < 0 || new_y < 0 || new_x >= grid[0].size() || new_y >= grid.size()) {
            continue;
        }

        if (grid[new_y][new_x] == '#') {
            continue;
        }

        if (visited.find(new_pos) != visited.end()) {
            continue;
        }
        
        the_max = max(the_max, recurse(grid, new_pos, end, visited, steps + 1));
    }

    visited.erase(pos);
    return the_max;
}


int main() {
    ifstream file("23.dat");
    string line;

    vector<vector<char>> grid;
    
    pair<int, int> end_pos, start_pos;
    
    int y = 0;
    while (getline(file, line)) {
        vector<char> new_row;

        for (int x = 0; x < line.length(); x++) {
            if (line[x] == '.') {
                if (y == 0) {
                    start_pos = make_pair(x, y);
                }
                end_pos = make_pair(x, y);
            }
            new_row.push_back(line[x]);
        }
        grid.push_back(new_row);
        y++;
    }

    unordered_set<pair<int, int>, pair_hash> the_set;
    cout << recurse(grid, start_pos, end_pos, the_set, 0) << endl;


    auto end = high_resolution_clock::now();
    auto duration = duration_cast<seconds>(end - start);
    cout << endl << duration.count() << " seconds taken" << endl;
    return 0;
}
